app/screens/convert_gm_to_channel/convert_gm_to_channel.tsx

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {useEffect, useRef, useState} from 'react';
+
+import {fetchChannelMemberships, getGroupMessageMembersCommonTeams} from '@actions/remote/channel';
+import {PER_PAGE_DEFAULT} from '@client/rest/constants';
+import {useServerUrl} from '@context/server';
+
+import ConvertGMToChannelForm from './convert_gm_to_channel_form';
+import {Loader} from './loader';
+
+import type UserProfile from '../user_profile/user_profile';
+
+type Props = {
+    channelId: string;
+    currentUserId?: string;
+}
+
+const loadingIndicatorTimeout = 1200;
+
+const ConvertGMToChannel = ({
+    channelId,
+    currentUserId,
+}: Props) => {
+    const [loadingAnimationTimeout, setLoadingAnimationTimeout] = useState(false);
+    const [commonTeamsFetched, setCommonTeamsFetched] = useState(false);
+    const [channelMembersFetched, setChannelMembersFetched] = useState(false);
+    const [commonTeams, setCommonTeams] = useState<Team[]>([]);
+    const [profiles, setProfiles] = useState<UserProfile[]>([]);
+
+    const serverUrl = useServerUrl();
+    const mounted = useRef(false);
+
+    const loadingAnimationTimeoutRef = useRef<NodeJS.Timeout>();
+
+    useEffect(() => {
+        loadingAnimationTimeoutRef.current = setTimeout(() => setLoadingAnimationTimeout(true), loadingIndicatorTimeout);
+        async function work() {
+            const {teams} = await getGroupMessageMembersCommonTeams(serverUrl, channelId);
+            if (!teams || !mounted.current) {
+                return;
+            }
+            setCommonTeams(teams);
+            setCommonTeamsFetched(true);
+        }
+
+        work();
+
+        return () => {
+            if (loadingAnimationTimeoutRef.current) {
+                clearTimeout(loadingAnimationTimeoutRef.current);
+            }
+        };
+    }, []);
+
+    const matchUserProfiles = (users: UserProfile[], members: ChannelMembership[]) => {
+        // Gotta make sure we use profiles that are in members.
+        // See comment in fetchChannelMemberships for more details.
+
+        const usersById: {[id: string]: UserProfile} = {};
+        users.forEach((profile) => {
+            if (profile.id !== currentUserId) {
+                usersById[profile.id] = profile;
+            }
+        });
+
+        const filteredUsers: UserProfile[] = [];
+        members.forEach((member) => {
+            if (usersById[member.user_id]) {
+                filteredUsers.push(usersById[member.user_id]);
+            }
+        });
+
+        return filteredUsers;
+    };
+
+    useEffect(() => {
+        mounted.current = true;
+
+        const options: GetUsersOptions = {sort: 'admin', active: true, per_page: PER_PAGE_DEFAULT};
+        fetchChannelMemberships(serverUrl, channelId, options, true).then(({users, members}) => {
```

Comment:
If there is no network connection, or MM server is down, the loader will continue to spin forever. Do we already have example of handling such case in mobile app? I'd queue this for next iteration, as in, not a blocking for now IMO.

Reply:
If the API call fails for some reason, we return an empty array of users, so no usernames in UI are displayed.

Reply:
Can you share how we handle network errors? Any example?

Reply:
In this particular case you are right. `fetchChannelMemberships` does not handle the errors, and just return empty list of users and members in case there is a failure.

I would just update `fetchChannelMemberships` so it considers the errors that may be returned by `fetchProfilesInChannel` and/or `fetchChannelMembersByIds`. If you go deeper in this functions, you will see how it is handled at that level. Long story short, the client function may throw an exception, and most of the fetch functions use a try catch to handle the throw.

If you mean about handling at a higher level (i.e. how to explain the user there has been an error) that is more a UX question, and I would say something to be handled case by case. Some things we want to fail silently, some things we want to show some message, but that is better to be answered by UX.

Reply:
Okay, if I think more about this, since failure to fetch channel memberships isn't something that would block user from converting the GM, I don't want to show error if we get error during that part. So with that, do you think current implementation is fine?

Reply:
I would have to see how the behavior goes (I haven't run the code yet).
From what I am reading, the problem may not be exactly here. It seems that if there is no internet connection, `getGroupMessageMembersCommonTeams` will return `undefined` for teams. Therefore, `commonTeamsFetched` will always be false, and `showLoader` always true. Therefore... we will have an infinite loading.

If we set `commonTeamsFetched` no matter the result of `getGroupMessageMembersCommonTeams`, the underlying component will say "I have 0 common teams, so I will say `there are no common teams`", which is not true. You just don't know because you are offline.

But all this is based on what I am reading, so I may be missing some important piece.

Reply:
Yes, what you are saying is absolutely correct. I'll add error handling for `getGroupMessageMembersCommonTeams` but I think we're fine with how it works if `fetchChannelMemberships` fails - infinite load. If not, I'll queue this up for second iteration as it shouldn't block merging of this PR.